nucleus/utils.py

```diff
@@ -422,3 +435,120 @@ def paginate_generator(
+    s3_bucket, s3_key = s3_path.split("//", 1)[-1].split("/", 1)
+    return s3_bucket, s3_key
+
+
+def fetch_image(s3_url: str):
+    s3_bucket, s3_key = split_s3_bucket_key(s3_url)
+    image = Image.open(
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_key).get()["Body"]
+    )
+    return image
+
+
+def fetch_chip(ref_id: str):
+    image_loc = None
+    annotation_loc = None
+    if ref_id.startswith("s3"):
+        s3_bucket, s3_key = split_s3_bucket_key(ref_id)
+        try:
+            boto3.resource("s3").Bucket(s3_bucket).Object(
+                s3_key + ".jpeg"
+            ).load()
+            image_loc = ref_id + ".jpeg"
+        except ClientError:
+            return None, None
+        try:
+            boto3.resource("s3").Bucket(s3_bucket).Object(
+                s3_key + ".json"
+            ).load()
+            annotation_loc = ref_id + ".json"
+        except ClientError:
+            return image_loc, None
+    else:
+        if os.path.exists(ref_id + ".jpeg"):
+            image_loc = ref_id + ".jpeg"
+            if os.path.exists(ref_id + ".json"):
+                annotation_loc = ref_id + ".json"
+    return image_loc, annotation_loc
+
+
+def write_chip(
+    ref_id: str, image: Image.Image, annotations: List[Dict[str, str]]
+):
+    if ref_id.startswith("s3"):
+        s3_bucket, s3_key = split_s3_bucket_key(ref_id)
+        byteio = io.BytesIO()
+        image.save(byteio, format="jpeg")
+        byteio.seek(0)
+        boto3.resource("s3").Bucket(s3_bucket).Object(
+            s3_key + ".jpeg"
+        ).upload_fileobj(byteio)
+        annotation_loc = None
+        if len(annotations) > 0:
+            boto3.resource("s3").Bucket(s3_bucket).Object(
+                s3_key + ".json"
+            ).put(
+                Body=json.dumps(annotations, ensure_ascii=False).encode(
+                    "UTF-8"
+                ),
+                ContentType="application/json",
+            )
+            annotation_loc = ref_id + ".json"
+        return ref_id + ".jpeg", annotation_loc
+    else:
+        dirs = ref_id.rsplit("/", 1)[0]
+        os.makedirs(dirs, exist_ok=True)
+        image_loc = ref_id + ".jpeg"
+        annotation_loc = None
+        image.save(image_loc)
+        if len(annotations) > 0:
+            annotation_loc = ref_id + ".json"
+            with open(annotation_loc, "w", encoding="utf-8") as f:
+                json.dump(annotations, f, ensure_ascii=False)
+        return image_loc, annotation_loc
+
+
+def generate_offsets(w: int, h: int, chip_size: int, stride_size: int):
+    xs = np.arange(0, w - stride_size, chip_size - stride_size)
+    ys = np.arange(0, h - stride_size, chip_size - stride_size)
+    if len(xs) > 1:
+        xs = np.round(xs * (w - chip_size) / xs[-1]).astype(int)
+    if len(ys) > 1:
+        ys = np.round(ys * (h - chip_size) / ys[-1]).astype(int)
+    yield from product(ys, xs)
+
+
+def chip_annotations(data, x0: int, y0: int, x1: int, y1: int):
```

Comment:
This needs a docstring explaining what's going on. Why are you creating annotations for each chip?